js/spiral.js

```diff
@@ -0,0 +1,529 @@
+            'id': 'spiral',
+            'name': 'Spiral',
+            'countries': [ 'HK' ], // Hong Kong
+            'version': 'v1',
+            'rateLimit': 500,
+            'has': {
+                'CORS': false,
+                'cancelOrders': true,
+                'fetchOHLCV': true,
+                'fetchMyTrades': true,
+                'fetchOrder': false,
+                'fetchOrders': true,
+                'fetchOpenOrders': true,
+                'fetchCurrencies': true,
+                'fetchTicker': false,
+            },
+            'timeframes': {
+                '1m': 1,
+                '5m': 5,
+                '15m': 15,
+                '30m': 30,
+                '1h': 60,
+                '2h': 2 * 60,
+                '4h': 4 * 60,
+                '6h': 6 * 60,
+                '12h': 12 * 60,
+                '1d': 24 * 60,
+                '1w': 7 * 24 * 60,
+            },
+            'urls': {
+                'logo': 'https://www.spiral.exchange/static/media/logo-dark.ebcfc5ba.svg',
+                'api': 'https://api.spiral.exchange',
+                'www': 'https://www.spiral.exchange',
+                'doc': 'https://docs.spiral.exchange',
+                'fees': 'https://support.spiral.exchange/hc/en-us/articles/360006390493-Trading-Fee',
+            },
+            'api': {
+                'public': {
+                    'get': [
+                        'currencies',
+                        'products',
+                        'klines',
+                        'orderbook',
+                        'trades',
+                    ],
+                },
+                'private': {
+                    'get': [
+                        'wallet/balances',
+                        'myTrades',
+                        'order',
+                    ],
+                    'post': [
+                        'order',
+                    ],
+                    'delete': [
+                        'order',
+                        'order/all',
+                    ],
+                },
+            },
+            'fees': {
+                'trading': {
+                    'tierBased': false,
+                    'percentage': true,
+                    'taker': 0.001,
+                    'maker': 0.002,
+                },
+            },
+            'options': {
+                'apiExpire': 5, // api auth expires in 5 sec,
+                'defaultCount': 10,
+            },
+        });
+    }
+
+    toMilliseconds (microSeconds) {
+        return Math.floor (microSeconds / 1000);
+    }
+
+    covertSide (side) {
+        if (side === 'bid') {
+            return 'buy';
+        } else if (side === 'ask') {
+            return 'sell';
+        }
+        return undefined;
+    }
+
+    async fetchMarkets (params = {}) {
+        let response = await this.publicGetProducts ();
+        let markets = this.safeValue (response, 'data', []);
```

Comment:
Can you do the same edit everywhere throughout the code:

```
response['trades'] || []
```

↓

```
this.safeValue (response, 'trades', [])
```

? This does work in JS, but if the key does not exist there, it will throw an exception in Python and PHP...

Reply:
thanks, I have check the codes and make the changes.